### DISCUSSION TITLE
How to use ValueConverter service for converting e.g. ID and STRING values

### DISCUSSION THREAD

**Discussion by hallvard** [eclipse-langium/langium#147] [Q&A] [answered]
I'm traversing the AST I receive from the parser, and can traverse using the generated isXYZ predicates and types. However, when I read a STRING feature, it includes the quotes, and although I can strip them off myself, I guess the correct way to do it is to use a ValueConverter. However, it has a convert methods that takes the value and a `CstNode` and I haven't found out how to obtain the the latter. My first guess was that this was related to my grammar access, but although I can get the RuleCall corresponding to the STRING feature, I don't know how to turn that into a CstNode.  I notice the AstNode as a $cst field and that inside there is a child corresponding to my STRING feature, but I cannot find API for getting it.

What is the correct approach?

**Comment by msujew** [comment, marked as answer]
Hi @hallvard, I guess we f'd something up. The default value converter will strip the quotes if you rename the `STRING` terminal to `string` (even though it will show a warning). The reason for that is this code:

https://github.com/langium/langium/blob/fbdef5d08b059e182a86db60dc06ef46e44d35e5/packages/langium/src/parser/value-converter.ts#L32-L36

On a more general note: You don't have to retrieve the `CstNode` yourself, you can just override the protected `runConverter` method, and bind your custom `ValueConverter` in your services declaration.

**Reply by msujew** [reply]
Did you rename every reference to the previous `STRING` terminal rule? Usually this error comes up when a rule could not be resolved.

We should really run validation of the input file in the `langium-cli`.

**Reply by hallvard** [reply]
No, I didn't, that's why I removed my answer ;-). Your suggestion seems to work.

**Reply by msujew** [reply]
Thought so ;)

Alright, good to know. We will probably make the default implementation of the `ValueConverter` case insensitive, since we want to support any terminal like `string`, `STRING` or `String` with the same conversion.

**Reply by spoenemann** [reply]
@msujew shall we rename our terminal rule to `STRING` as well? That would be more consistent (and eliminate the validation warning ;-) )

**Reply by hallvard** [reply]
How would I define a 'datatype' rule for boolean with a ValueConverter translating yes/no to true/false and 'wire' it into the services object?

**Reply by spoenemann** [reply]
Just use `boolean` as the return type. The default value converter will take care of that:

https://github.com/langium/langium/blob/fbdef5d08b059e182a86db60dc06ef46e44d35e5/packages/langium/src/parser/value-converter.ts#L37-L41

**Reply by msujew** [reply]
Actually, the default value converter does not support transforming `yes/no` to `true/false`. @hallvard you are right, you would have to define your own value converter for that like this:

```ts
export class QuizValueConverter extends DefaultValueConverter {
    protected runConverter(rule: GrammarAST.AbstractRule, input: string, cstNode: CstNode): ValueType {
        if (rule.name === 'YesNo') {
            return input === 'yes';
        } else {
            return super.runConverter(rule, input, cstNode);
        }
    }
}
```
Then, bind this new value converter in your services like this:

```ts
export const QuizModule: Module<QuizServices, PartialLangiumServices & QuizAddedServices> = {
    // The rest of the module comes here
    parser: {
        ValueConverter: () => new QuizValueConverter()
    },
```

**Reply by drhagen** [reply]
@msujew Would you mind updating this solution? `Rule` does not appear to be defined anymore, and VSCode is giving me red squiggles on `parser`.

**Reply by msujew** [reply]
I've updated the `Rule` type reference. What kind of error do you get on `parser`? That part at least hasn't changed in the meantime.

**Reply by drhagen** [reply]
Never mind, `parser` is fine now. Thanks for updating the example.
